### PR TITLE
refactor(common): generate more code on system params with macro

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

When I was working on #14528, I find a lot of places have to be updated when trying to introduce a new system parameter. This PR is an attempt to improve that.

- Generate `SystemConfig` and initial system params.
- Generate `SystemParamsRead` trait, so developers will be reminded to add a new method on `SystemParamsReader` after introducing a new system param.
- Generate unwrapped default value functions, as suggested by @xxchan in https://github.com/risingwavelabs/risingwave/pull/14528#discussion_r1449908549.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/config.rs [hunk 1]
```diff
@@ -849,65 +850,27 @@ pub struct BatchDeveloperConfig {
...
+    ($({ $field:ident, $type:ty, $default:expr, $is_mutable:expr, $doc:literal, $($rest:tt)* },)*) => {
+        paste::paste!(
+            /// The section `[system]` in `risingwave.toml`. All these fields are used to initialize the system
+            /// parameters persisted in Meta store. Most fields are for testing purpose only and should not be
+            /// documented.
+            #[derive(Clone, Debug, Serialize, Deserialize, DefaultFromSerde)]
+            pub struct SystemConfig {
+                $(
+                    #[doc = $doc]
+                    #[serde(default = "default::system::" $field "_opt")]
```
Comment (xxchan) [inline src/common/src/config.rs:864]: Oh here is `default_opt` used

**Review by yezizp2012 — APPROVED** [review]
LGTM!

File: src/common/src/system_param/mod.rs [hunk 2]
```diff
@@ -83,23 +110,42 @@ macro_rules! def_key {
...
+
+/// Define default value functions for those with `Some` default values.
+macro_rules! def_default {
+    ($({ $field:ident, $type:ty, $default: expr, $($rest:tt)* },)*) => {
+        $(
+            def_default!(@ $field, $type, $default);
+        )*
+    };
+    (@ $field:ident, $type:ty, None) => {};
+    (@ $field:ident, $type:ty, $default: expr) => {
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/common/src/system_param/mod.rs:134]: Where is this non-opt version used?

File: src/common/src/system_param/reader.rs [hunk 3]
```diff
@@ -31,67 +50,71 @@ impl From<PbSystemParams> for SystemParamsReader {
     }
 }
 
-impl SystemParamsReader {
-    pub fn barrier_interval_ms(&self) -> u32 {
+// Unwrap the field if it exists from the initial release.
+// Otherwise, fill it with the backward compatibility logic.
```
Comment (xxchan) [inline src/common/src/system_param/reader.rs:54]: What is the "release"? and what is compat logic?

Comment (xxchan) [reply src/common/src/system_param/reader.rs:54]: I kind of get it, but still confused. e.g., why can we unwarp `wasm_storage_url`

File: src/common/src/system_param/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xxchan) [reply src/common/src/system_param/mod.rs:134]: In the Reader Impl

File: src/common/src/system_param/reader.rs (hunk 3 above)
Comment (BugenZhao) [reply src/common/src/system_param/reader.rs:54]: It must be a bug. 😄

Comment (xxchan) [reply src/common/src/system_param/reader.rs]: ... and why can we unwrap `state_store`, which defaults to `None`

Comment (xxchan) [reply src/common/src/system_param/reader.rs]: It's not related with the `Option` in prost at all.

Comment (BugenZhao) [reply src/common/src/system_param/reader.rs]: It's okay. 😄 It'll be checked on startup.

https://github.com/risingwavelabs/risingwave/blob/ffd922ff26ab4a374a08e09178b2a5e6f3f8e135/src/meta/src/manager/system_param/mod.rs#L59-L70

**Pull request merged**